File: src/leetcode/python/medium/BinaryTreeRightSideView_199.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreeRightSideView_199(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(node, depth):
            if node:
                if len(rsw) == depth:
                    rsw.append(node.val)
                dfs(node.right, depth + 1)
                dfs(node.left, depth + 1)

        rsw = []
        dfs(root, 0)
        return rsw

File: src/leetcode/java/BinaryTreeRightSideView_199.java
package leetcode.java;

import java.util.LinkedList;
import java.util.List;

public class BinaryTreeRightSideView_199 {
  private List<Integer> list;
  private int depth;
  private int maxDepth;

  public List<Integer> rightSideView(TreeNode root) {
    list = new LinkedList<>();
    maxDepth = depth = 0;
    if (root == null) return list;
    list.add(depth, root.val);
    dfs(root.left);
    dfs(root.right);
    return list;
  }

  private void dfs(TreeNode node) {
    if (node == null) return;
    depth++;
    if (maxDepth < depth) {
      maxDepth = depth;
      list.add(depth, node.val);
    } else {
      list.set(depth, node.val);
    }
    dfs(node.left);
    dfs(node.right);
    depth--;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
